### ISSUE TITLE
Add Native (non-WSL) Windows support to semgrep

### ISSUE THREAD

**Issue by mschwager** [semgrep/semgrep#1330]
Creating this issue to begin tracking requests for Semgrep support on Windows. Note that we expect Semgrep to work via `pip` installation on WSL, so this issue is focused on native Windows support. If you need _native_ Windows support please chime in here :+1: 

Depends on:
- [x] #2956
- [ ] #2991
- [x] #3061
- [ ] [issue needed] Eliminate calls to fork(), which requires parser to run in the same semgrep process, but has a memory leak as a workaround for a segfault that would need to be fixed ([segfault issue](https://github.com/returntocorp/ocaml-tree-sitter-semgrep/issues/65)). The memory leak may be acceptable for small scans.
- [x] [issue needed] Set up Windows testing in CI

**Comment by parsiya** [comment]
Attaching the error after running `pip3 install semgrep` on Windows 10 (1909) and Python 3.8.1. Seems to be a similar issue to #1336 but not the path separator problem.

I am running it via WSL2 so it's not a deal breaker but it would be nice to have Windows support for the VS Code extension (unless I can figure out how to run it via WSL2).

The log is attached but the important part appears to be:

```
error: can't copy 'C:\Users\phakimian\AppData\Local\Temp\1\pip-install-r20usipa\semgrep-files\semgrep-core': doesn't exist or not a regular file
```
Looking at the log, I cannot see the file being copied. 

Unrelated: There's also this warning so something to look out for in the future.

```
DEPRECATION: Could not build wheels for semgrep which do not use PEP 517. pip will fall back to legacy
'setup.py install' for these. pip 21.0 will remove support for this functionality. A possible replacement is to
fix the wheel build issue reported above. You can find discussion regarding this at
https://github.com/pypa/pip/issues/8368.
```

[pip-install-semgrep-windows.txt](https://github.com/returntocorp/semgrep/files/5070821/pip-install-semgrep-windows.txt)

**Comment by bobby-lin** [comment]
Bumping this up. Is this work in progress?

**Comment by zain-CSsol** [comment]
Yes i think it's still in progress i am also backtracking the issue for solution, anyhow if any one of you came to know the solution then do let me know

**Comment by DrewDennison** [comment]
related to #2252 and #2565 we would need to do cross-platform builds as part of our release process

**Comment by DrewDennison** [comment]
marking as low priority given WSL and docker workarounds

**Comment by fasiha** [comment]
We're interested in running semgrep on Windows natively (Docker and WSL won't work for us unfortunately). I've been working with [OCaml for Windows](https://fdopen.github.io/opam-repository-mingw/) which, with a couple of changes to a couple of build scripts, seems to be working well.

I ran into this though:
```
opam install -y --deps-only ./semgrep-core
The following dependencies couldn't be met:
  - semgrep → parmap
      unmet availability conditions, e.g. os != "win32"

No solution found, exiting
```
It looks like [parmap](https://github.com/rdicosmo/parmap/issues/76)'s `fork`-based design won't work natively in Windows. I'm alas not fluent in OCaml (I want to be!) but I'm curious in your estimation how easy it would be to replace parmap with a cross-platform alternative, perhaps [procord](https://github.com/cryptosense/procord) mentioned in that parmap issue?

**Comment by mschwager** [comment]
Hey @fasiha,

Thanks for looking into this! If you're interested in continuing to push this forward, then we could take a look at the changes. Supporting Windows is an additional maintenance burden for us, but it does open the door for Semgrep to run on many many more systems. There's no guarantee that we could take on that maintenance right now, but if the changes are fairly minimal, and we can leverage CI to handle the build/test/release process, then we're definitely open to the idea.

Regarding parmap, I'm totally ignorant with all things OCaml, but cc @aryx @mjambon who may be able to help.

**Comment by mschwager** [comment]
To add a bit more color to the overall objective, I think, by far, the biggest hurdle to running Semgrep on Windows is getting the OCaml binaries compiled and released. Our Python code should currently work on Windows as-is. Of course, there may be a few bugs here and there, but overall I think it should Just Work.

Semgrep installation can be forced on Windows if you'd like to test things out - you'll just need to bring your own `semgrep-core` and `spacegrep` (notice a trend? :)). The following command should do the trick:

```
$ SEMGREP_SKIP_BIN=true SEMGREP_FORCE_INSTALL=true python -m pip install semgrep
```

*Note: this should install the source package (`.tar.gz`) on Windows, not the wheel package (`.whl`).*

This will install the Python code as `pip` normally does. From here, Semgrep will look for `semgrep-core` and `spacegrep` according to [`shutil.which`](https://docs.python.org/3/library/shutil.html#shutil.which). My Windows foo is pretty weak, but it looks like this means it will respect the `PATHEXT` environment variable, so if you place the OCaml binaries somewhere in that path it should work :+1:

**Comment by fasiha** [comment]
Thanks for the detailed notes @mschwager!

> the biggest hurdle to running Semgrep on Windows is getting the OCaml binaries compiled and released

Thanks for confirming, this matches my expectation. Here's how I got to it:

1. Download [Win10 evaluation images](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/) from Microsoft.
2. Install [OCaml for Windows](https://fdopen.github.io/opam-repository-mingw/) which is—this is important—a *clone* of Cygwin with *all-native* versions of opam, flexdll and aspcud (no Cygwin.dll dependencies)
3. Download [Cygwin setup-x86_64.exe](https://cygwin.com/install.html) and run it to install additional dependencies:
    - binutils (for `ar`)
    - pkg-config
    - git? (I installed Git for Windows outside of this but it's probably more convenient to install it here)
4. In a Cygwin Bash, `opam switch create 4.10.2+mingw64c`
5. Clone Semgrep, check out v0.46.0 tag.
6. At this point `opam exec -- make setup` will fail, so I had to
7. Run `PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:$PKG_CONFIG_PATH` because for some reason, pkg-config won't see tree-sitter otherwise
8. In [ocaml-tree-sitter/scripts/install-tree-sitter-lib line 73](https://github.com/returntocorp/ocaml-tree-sitter/blob/8f2515c1ccbacff797c8d884379f8219658cc5fd/scripts/install-tree-sitter-lib#L73): remove `sudo`
9. In [ocaml-tree-sitter/scripts/configure-tree-sitter line 20](https://github.com/returntocorp/ocaml-tree-sitter/blob/8f2515c1ccbacff797c8d884379f8219658cc5fd/scripts/configure-tree-sitter#L20) change the pkg-config to `pkg-config --libs-only-L tree-sitter` (there's no ldconfig in Cygwin so pkg-config without flags won't find tree-sitter, but with `--libs-only` or `--cflags-only` will work fine)
10. At this point, I hit the `parmap` issue.

We are probably ok with doing this manually for now, though we would need help with `parmap`, but having written all this out, I think all of this so far can readily be scripted, including the extra Cygwin support packages—just a matter of time, which I totally understand is in short supply :).

Does this help? I don't think I missed any steps but happy to add extra details. I'm not sure, if there's a workaround to the `parmap` issue, how many more issues there will be to get spacegrep and semgrep-core binaries but if not, that would be really cool!

**Comment by mschwager** [comment]
This is perfect!

Moving forward here would look something like:

- Gather all the commands necessary to build these binaries from scratch on Windows
- Create a new Github Action workflow for building on Windows, similar to [Linux and OSX](https://github.com/returntocorp/semgrep/blob/develop/.github/workflows/build-tests.yml)
- Ensure the binaries are properly tested and packaged in the Windows workflow environment
- Update `setup.py` to handle Windows environments
- Release Windows support!

All your obstacles seem fixable. Regarding (8), I would say we generally shouldn't rely on `sudo`, and instead should set up the correct permissions such that we don't need it. (9) seems like an environmental thing that shouldn't be a major problem to handle on a per-environment basis. And (10) seems like we have a workaround too (`procod`, or simply don't leverage `parmap` in Windows environments).

I think we'll need some help from our OCaml experts, but this all sounds reasonable to me :+1:

**Comment by mjambon** [comment]
Parmap uses `fork()`. It's used by the `-j` option of semgrep-core for processing multiple input files in parallel. If we disable it, perhaps by installing a dummy parmap library when on Windows, we could then get semgrep-core to build and run on Windows. I don't know if this option is even used by the python wrapper at the moment or what the python wrapper uses for running tasks in parallel.

Another issue is that we avoid [this segfault turned into a memory leak](https://github.com/returntocorp/ocaml-tree-sitter/issues/65) by running parsing jobs in a detached process, which also uses `fork()`. We could eliminate this need by fixing the bug.

We have various bash scripts as well, which are used only at build time or testing time. There must be also other small issues like unexpected `.exe` extensions.

Off the top of my head, that's all I can think of.

And `sudo` is causing troubles to a lot of people. It's used to install libtree-sitter. I don't know how to do otherwise and some help would be appreciated. See my [recent comment here](https://github.com/conda-forge/staged-recipes/pull/14494#issuecomment-821579502). Edit: [dedicated ticket for sudo is here](https://github.com/returntocorp/semgrep/issues/2956).

**Comment by fasiha** [comment]
Thanks for the notes @mjambon!!

I am placing myself entirely in your hands for both the `parmap` and segfault-vs-mem-leak issues ☺️ but, from my perspective (which is: first Windows build, then cross-platform conda build), the sudo isn't a big deal because we can patch that :) I'm testing that right now on the conda Linux CI.

And, because OCaml for Windows is a Cygwin environment, those Bash scripts should be fine 🤞. I just had to dig around to see, for example, which Cygwin package would give me `ar` (turned out to be binutils, same as Debian…), hopefully any other Bash issues just are as straightforward.

**Comment by fasiha** [comment]
Some more information and a request for help 🙏!

Since I couldn't get farther with a native Windows build due to `parmap` unavailable on MinGW (which is what OCaml for Windows uses to produce native stand-alone Windows builds), I switched gears to see if I could use full Cygwin to get a build that depends on cygwin1.dll. I understand  this might not be a long-term solution for, e.g., conda, but would help us out, plus it might help identify other issues that we will face when creating a full native build.

Following up from my [steps 1-10 above](https://github.com/returntocorp/semgrep/issues/1330#issuecomment-821246324)—after steps 1-10 above, I did the following:

11. Run Cygwin's setup_x86_64.exe to add libpcre-devel. With this, `make setup` works!
12. Following the [Homebrew recipe](https://github.com/Homebrew/homebrew-core/blob/master/Formula/semgrep.rb) I was able to build spacegrep and ocaml-tree-sitter!
13. I ran into a problem when I ran `opam exec -- make all` in semgrep-core. Many files in `~/.opam/default/lib/spacegrep/dune-package` were invalid: they were supposed to be symlinks but, whatever tool created them there left them broken because they were flat files containing text, e.g., `!<symlink>▒▒../../../../default/spacegrep.dune-package`. I wrote a quick shell script to find all these broken "symlinks" and overwrite them from the versions under `spacegrep/_build/default/`.
14. Then, I hit another roadblock: `opam exec -- make all` prints an error: "Error on dynamically loaded library: dllctypes_stubs.so: No such file or directory". ***Edit*** I see this file in ~/.opam/ocaml-base-compiler.4.10.2/lib/ctypes/dllctypes_stubs.so so I run `export CAML_LD_LIBRARY_PATH=$CAML_LD_LIBRARY_PATH:$HOME/.opam/ocaml-base-compiler.4.10.2/lib/ctypes` ([tip](https://discuss.ocaml.org/t/dune-exec-c-stubs-and-caml-ld-library-path/2872)).
    - It seemed like `opam exec` was ignoring this flag? But `dune build` worked fine and then I finished up with `make all && make install`. 
15. Final step, `./_build/install/default/bin/semgrep-core` is one of those fake symlinks again, so copy `./_build/default/src/cli/Main.exe` to `semgrep-core`.

Success! We have `semgrep-core` and `spacegrep` binaries 🥳!
- spacegrep: `spacegrep/_build/default/src/bin/Space_main.exe`
- semgrep-core: `semgrep-core/_build/default/src/cli/Main.exe`

Next are my steps for getting semgrep to work via the Python installer. I did this on *another* virtual machine to simulate a new computer that didn't build it.

1. Download and extract the Semgrep tarball from PyPI. Then, in Command Prompt cmd.exe,
2. `set SEMGREP_FORCE_INSTALL 1` (otherwise installer bails because it sees Windows)
3. `set SEMGREP_CORE_BIN=z:\semgrep-core`
4. `set SPACEGREP_BIN=z:\spacegrep`
5. `python -m pip install .`
6. This copies the `semgrep` executable to `C:\Users\User\AppData\Local\Programs\Python\Python39\Scripts` but we need to copy four DLLs from Cygwin to the same path to sit alongside semgrep executable:
    - `cygwin1.dll`
    - `cygstdc++-6.dll`
    - `cygpcre-1.dll`
    - `cyggcc_s-seh-1.dll`

At this stage, when I attempt to invoke semgrep, I run into an unusual problem:
```
C:\foo>semgrep -e "eval(...)" --lang=py demo.py
an internal error occured while invoking semgrep-core while running rule '-'. Consider skipping this rule and reporting this issue.
        unknown exception: End_of_file
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep


C:\foo>semgrep -e "$X == $X" --lang=py demo.py
an internal error occured while invoking semgrep-core while running rule '-'. Consider skipping this rule and reporting this issue.
        unknown exception: End_of_file
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep


C:\foo>semgrep --config=rules.yaml --lang=py foo.py
running 44 rules...
  0%|                                                                                                             |0/44
an internal error occured while invoking semgrep-core while running rule 'testmodule.Test'. Consider skipping this rule and reporting this issue.
        unknown exception: End_of_file
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep
```

I can't seem to get any output from semgrep?

Note that `foo.py` and `demo.py` both exist: if I attempt to run the above commands with a non-existent file, semgrep correctly says "File not found" (and then "ran 1 rules on 0 files: 0 findings").

I ran `unix2dos` on my `rules.yaml` file in case that was causing the EOF exception from semgrep, but that doesn't seem to be the issue since `-e` rules don't work. Just in case, I also ran these on my Python test files but that doesn't help either.

I don't think it's the quote convention on cmd.exe since `semgrep -e x --lang=py demo.py` results in the same EOF unknown exception.

`core_runner.py` says `See format_output_exception in semgrep O'Caml for details on schema` but I don't see this function in the OCaml semgrep repo? I do see lots of use of `End_of_file`, maybe that'll help me figure out what the issue might be—obviously it works fine in Unix, something Windows-specific (or maybe Cygwin-specific) might be biting us here.

I'll continue to poke around but wanted to update you 😄!

**Comment by fasiha** [comment]
Sorry to spam everyone on the thread again 🙇. I updated the [comment above](https://github.com/returntocorp/semgrep/issues/1330#issuecomment-823614219) because we found a fix for the dllctypes_stub.so issue and were able to get binaries on Cygwin (!!!), but then ran into some unexpected exceptions when running Semgrep from the Python installation.

Any tips would be most appreciated! Thank you all for your support!!!

**Comment by mschwager** [comment]
> I updated the comment above because we found a fix for the dllctypes_stub.so issue and were able to get binaries on Cygwin

Amazing! Nice work!

The only `End_of_file` I see in `semgrep-core` is:

```ocaml
(* mostly a copy of Parse_info.full_charpos_to_pos_large *)
let line_col_to_pos = fun file ->

  let chan = open_in file in
  let size = Common2.filesize file + 2 in

  let charpos   = ref 0 in
  let line  = ref 0 in
  let h = Hashtbl.create size in

  let full_charpos_to_pos_aux () =
    try
      while true do begin
        let s = (input_line chan) in
        incr line;

        (* '... +1 do'  cos input_line dont return the trailing \n *)
        for i = 0 to (String.length s - 1) + 1 do
          Hashtbl.add h (!line, i) (!charpos + i);
        done;
        charpos := !charpos + String.length s + 1;
      end done
    with End_of_file ->
      (* bugfix: this is wrong:  Hashtbl.add h (!line, 0) !charpos;
       * because an ident on the last line would get
       * the last charpos.
      *)
      ()
  in
  full_charpos_to_pos_aux ();
  close_in chan;
  h
```

The most curious line to me is: `(* '... +1 do'  cos input_line dont return the trailing \n *)`.

Could this be a Windows vs. Unix line ending issue? E.g. `\r\n` vs. `\n`?

Or maybe somehow related to https://github.com/returntocorp/semgrep/issues/2237?

cc @aryx who's more familiar with this code.

**Comment by fasiha** [comment]
> maybe somehow related to #2237?

Ohh, I wonder if it’s something related to shell ARG_MAX or max file name lengths or any number of Windows vs Unix differences, let me try that angle!

I’m also wondering if it’s possible for me to share this Windows image containing everything so you could poke around—it’s a Parallels image downloaded from Microsoft a week or two ago so it should be good for another couple of weeks (these expire after 30 days I believe). Let me know if anyone is interested in that, to save you from the steps above.

**Comment by mjambon** [comment]
> 13. I ran into a problem when I ran `opam exec -- make all` in semgrep-core. Many files in `~/.opam/default/lib/spacegrep/dune-package` were invalid: they were supposed to be symlinks but, whatever tool created them there left them broken because they were flat files containing text, e.g., `!<symlink>▒▒../../../../default/spacegrep.dune-package`. I wrote a quick shell script to find all these broken "symlinks" and overwrite them from the versions under `spacegrep/_build/default/`.

I don't see symlinks for these files on my ubuntu/linux machine, just regular files. I'm on dune 2.8.2. The installation of the spacegrep files into .opam is done with `dune install`, which is why I think it's an issue with dune.

> 15. Final step, `./_build/install/default/bin/semgrep-core` is one of those fake symlinks again, so copy `./_build/default/src/cli/Main.exe` to `semgrep-core`.

The `_build` folder is managed exclusively by dune, so yes, it seems to be an issue with some version of dune on Windows.

**Comment by fasiha** [comment]
> I don't see symlinks for these files on my ubuntu/linux machine, just regular files. I'm on dune 2.8.2. The installation of the spacegrep files into .opam is done with `dune install`, which is why I think it's an issue with dune.

I wonder if this is related to this argument in `dune build --help` (dune 2.8.5):
```
       --sandbox=VAL (absent DUNE_SANDBOX env)
           Sandboxing mode to use by default. Some actions require a certain
           sandboxing mode, so they will ignore this setting. The allowed
           values are: none, symlink, copy.
```
You most likely ran `opam init` with sandbox enabled, and I wonder if that either sets `DUNE_SANDBOX` or allows your install to copy the binaries rather than trying to symlink them in this way. I can't seem to get dune to `copy` these though: even in the `_build` directory, these are "symlinks", in quotes because they don't work in Cygwin and might remain a problem in native Windows build.

**Comment by mjambon** [comment]
@fasiha this seems tricky enough to ask directly the dune folks at https://github.com/ocaml/dune/issues. It's probably also worth probing discuss.ocaml.org since other users may have had to solve this.

**Comment by clintgibler** [comment]
This worked for someone on Windows:

> Using poetry as my package manager instead of pip in a test project on Windows 10 worked for me:

```
poetry init
poetry shell
poetry add semgrep
```

**Comment by AsquaredWsquared** [comment]
Sorry @clintgibler that was me giving you rubbish information (thanks again for the talk on Semgrep). It installs and can access the help but if I run the below in a poetry shell it isn't happy:

C:\Users\ilike\semgrep_test>semgrep --config=p/ci test.py    
using config from https://semgrep.dev/p/ci. Visit https://semgrep.dev/registry to see all public rules.
downloading config...
running 135 rules...
 34%|█████████████████████████████████████▍                                                                        |46/135
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\ilike\AppData\Local\pypoetry\Cache\virtualenvs\semgrep-test-5m_MXkUu-py3.9\Scripts\semgrep.exe\__main__.py", line 7, in <module>
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\__main__.py", line 20, in main
    cli()
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\cli.py", line 517, in cli
    semgrep.semgrep_main.main(
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\semgrep_main.py", line 270, in main
    ) = CoreRunner(
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\core_runner.py", line 778, in invoke_semgrep
    ) = runner_fxn(rules, target_manager, profiler)
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\core_runner.py", line 710, in _run_rules_direct_to_semgrep_core
    core_run = sub_run(
  File "c:\users\ilike\appdata\local\pypoetry\cache\virtualenvs\semgrep-test-5m_mxkuu-py3.9\lib\site-packages\semgrep\util.py", line 154, in sub_run
    result = subprocess.run(cmd, **kwargs)  # nosem: python.lang.security.audit.dangerous-subprocess-use.dangerous-subprocess-use
  File "C:\Program Files\Python39\lib\subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python39\lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python39\lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
OSError: [WinError 193] %1 is not a valid Win32 application``

**Comment by csrichter** [comment]
i'm also interested in in running semgrep on windows without WSL

i have tried installing semgrep via pip natively and that works, but running it leads to the error : ModuleNotFoundError: No module named 'resource'

i've read that the package "resource" is not available on windows

**Comment by aryx** [comment]
This is becoming higher priority

**Comment by aryx** [comment]
good progress this week: https://github.com/semgrep/semgrep/pull/9291
We manage to get semgrep-core and semgrep scan ci working natively under windows.
There is still lots of rough edges and stuff to fix, but this is promising.

**Comment by ievans** [comment]
@tpetr has your work closed this issue?

**Comment by L93DC** [comment]
When attempting to install llama-recipes on a Windows environment using the command provided by Llama (pip install llama-recipes), the installation process fails. The root cause appears to be the semgrep dependency, which currently does not support Windows.
Error Details:
During the installation, the following error is encountered:
Collecting semgrep>1.68 (from codeshield->llama-recipes)
  Using cached semgrep-1.101.0.tar.gz (27.6 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      Traceback (most recent call last):
        File "I:\Python\Examples\LlamaGuardCombat\.LlamaRecipesVenv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "I:\Python\Examples\LlamaGuardCombat\.LlamaRecipesVenv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "I:\Python\Examples\LlamaGuardCombat\.LlamaRecipesVenv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\qq202\AppData\Local\Temp\pip-build-env-17typjpg\overlay\Lib\site-packages\setuptools\build_meta.py", line 334, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\qq202\AppData\Local\Temp\pip-build-env-17typjpg\overlay\Lib\site-packages\setuptools\build_meta.py", line 304, in _get_build_requires
          self.run_setup()
        File "C:\Users\qq202\AppData\Local\Temp\pip-build-env-17typjpg\overlay\Lib\site-packages\setuptools\build_meta.py", line 320, in run_setup
          exec(code, locals())
        File "<string>", line 63, in <module>
      Exception: Semgrep does not support Windows yet, please try again with WSL or visit the following for more information: https://github.com/semgrep/semgrep/issues/1330
      [end of output]

note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

Environment:

Operating System: Windows 10/11
Python Version: [e.g., Python 3.9]
pip Version: [e.g., pip 23.0]
Virtual Environment: Yes/No (specify if applicable)

**Comment by milanwilliams** [comment]
👋 Good news, we're aiming to have native Windows support available for Semgrep by the end of July. 

Let me know if you'd like to be an early beta user! 🚀

**Comment by 1nhann** [comment]
> 👋 Good news, we're aiming to have native Windows support available for Semgrep by the end of July.
> 
> Let me know if you'd like to be an early beta user! 🚀

Will the Pro engine be released on Windows?

**Comment by nmote** [comment]
@1nhann Yes, it will.

**Comment by nmote** [comment]
Semgrep on Windows is now in public beta! See our [quickstart guide](https://semgrep.dev/docs/getting-started/quickstart) for installation instructions.
